source/s3_default_meta_request.c

```diff
@@ -254,7 +254,7 @@ static void s_s3_meta_request_default_request_finished(
     struct aws_s3_meta_request_default *meta_request_default = meta_request->impl;
     AWS_PRECONDITION(meta_request_default);
 
-    if (error_code == AWS_ERROR_SUCCESS && meta_request->headers_callback != NULL &&
+    if (meta_request->headers_callback != NULL && request->send_data.response_headers != NULL &&
```

Comment:
- At what case, the response headers will be null?
- What about get_meta_request/put_meta_request? Do those have the same issue?

Reply:
Thanks for looking! 

1) The response headers will be null for errors that prevent the HTTP request to be executed (e.g. error codes like `AWS_IO_DNS_INVALID_NAME`). In this case, the headers callback won't be invoked.
2) Get/Put meta requests are composed of sub-requests, and each sub-request could have distinct errors. Still evaluating how granular the error reporting of the parent meta request needs to be. I suggest handling them in a separate PR unless it could impact our decision to call or not to call the headers callback on errors for `Default` meta requests. Any thoughts?

Reply:
1. We should add some documentations [here](https://github.com/awslabs/aws-c-s3/blob/main/include/aws/s3/s3_client.h#L35-L62) or [here](https://github.com/awslabs/aws-c-s3/blob/main/include/aws/s3/s3_client.h#L142-L152) like what we did for [http](https://github.com/awslabs/aws-c-http/blob/main/include/aws/http/request_response.h#L130-L188), which will help others to understand how to callbacks will be invoked. 
2. I would like to have a consistent behavior for the meta requests. It may be not too bad to add the handle for get/put in this PR(This PR is relative small). I think we offer the header callback when error happens is a right thing. The put/get will also have a combined header to be offered to the user.

Reply:
Thank you for the references! I'll address both issues in this PR.